### ISSUE TITLE
GnuCash 3.8 Flatpak + Postgres 12.2 -> "Cannot save to database"

### ISSUE THREAD

**Issue by rob-hills** [flathub/org.gnucash.GnuCash#30]
I have been running GnuCash 2.6.19 on Ubuntu 18.04 LTS with Postgres 9.x as the back end for some time now and it has all been operating quite well.

I have recently decided to try the newest versions of GnuCash (3.8) and Postgres (12.2) together, and these are not working well together at all.  To get the newest version of GnuCash, I installed the Flatpak version.  I backed up my GnuCash database from Postgres 9.x and restored into version 12.2 without a problem.

GnuCash 3.8 starts up OK.  It now prompts me with a DB Login prompt each time I start but that's a minor issue at present.  My main problem is that it is only possible to save some types of entries in GnuCash.  Attempts to save anything else give database errors.

For example, I can create a new invoice.  I can add line items to my invoice - with each line I get a "Cannot save to database" error, but the line item data is all saved anyway!  However, if I try and post the invoice, I can fill out the "post invoice" form (labelled "question") OK, but when I click the OK button on that form, I get several "Cannot save to database" error dialogs and then GnuCash shuts down.

If I run GnuCash from the terminal, this is what I see during the above evolution:
```
flatpak run org.gnucash.GnuCash
Found Finance::Quote version 1.49.
**
gnc.engine:ERROR:libgnucash/engine/gnc-lot.c:497:gnc_lot_get_balance: assertion failed: (gnc_numeric_check (baln) == GNC_ERROR_OK)
Bail out! gnc.engine:ERROR:libgnucash/engine/gnc-lot.c:497:gnc_lot_get_balance: assertion failed: (gnc_numeric_check (baln) == GNC_ERROR_OK)
```

And this is the database log from Postgres:
```
2020-02-29 12:09:09.619 AWST [22127] gnucash@gnucash_rob_professional ERROR:  duplicate key value violates unique constraint "slots_new_pkey"
2020-02-29 12:09:09.619 AWST [22127] gnucash@gnucash_rob_professional DETAIL:  Key (id)=(105) already exists.
2020-02-29 12:09:09.619 AWST [22127] gnucash@gnucash_rob_professional STATEMENT:  INSERT INTO slots(obj_guid,name,slot_type,int64_val,timespec_val,numeric_val_num,numeric_val_denom,gdate_val) VALUES('828aab6c5d56428e99ab923bced3d152','date-posted',10,0,'1970-01-01 00:00:00',0,1,'20200225')
2020-02-29 12:25:46.718 AWST [23229] gnucash@gnucash_rob_professional ERROR:  duplicate key value violates unique constraint "slots_new_pkey"
2020-02-29 12:25:46.718 AWST [23229] gnucash@gnucash_rob_professional DETAIL:  Key (id)=(106) already exists.
2020-02-29 12:25:46.718 AWST [23229] gnucash@gnucash_rob_professional STATEMENT:  INSERT INTO slots(obj_guid,name,slot_type,int64_val,timespec_val,numeric_val_num,numeric_val_denom) VALUES('d8a5fbd764624b128f2c84ae3c89d67c','credit-note',1,0,'1970-01-01 00:00:00',0,1)
2020-02-29 12:26:51.420 AWST [23229] gnucash@gnucash_rob_professional ERROR:  duplicate key value violates unique constraint "slots_new_pkey"
2020-02-29 12:26:51.420 AWST [23229] gnucash@gnucash_rob_professional DETAIL:  Key (id)=(107) already exists.
2020-02-29 12:26:51.420 AWST [23229] gnucash@gnucash_rob_professional STATEMENT:  INSERT INTO slots(obj_guid,name,slot_type,int64_val,timespec_val,numeric_val_num,numeric_val_denom) VALUES('d8a5fbd764624b128f2c84ae3c89d67c','credit-note',1,0,'1970-01-01 00:00:00',0,1)
2020-02-29 12:27:33.124 AWST [23229] gnucash@gnucash_rob_professional ERROR:  duplicate key value violates unique constraint "slots_new_pkey"
2020-02-29 12:27:33.124 AWST [23229] gnucash@gnucash_rob_professional DETAIL:  Key (id)=(108) already exists.
2020-02-29 12:27:33.124 AWST [23229] gnucash@gnucash_rob_professional STATEMENT:  INSERT INTO slots(obj_guid,name,slot_type,int64_val,timespec_val,guid_val,numeric_val_num,numeric_val_denom) VALUES('7a0fd3d46a3e4a2f911ec324e5f8ea29','gncInvoice',9,0,'1970-01-01 00:00:00','d253666255b54b0b832538fd038b5293',0,1)
2020-02-29 12:28:07.862 AWST [23229] gnucash@gnucash_rob_professional ERROR:  duplicate key value violates unique constraint "slots_new_pkey"
2020-02-29 12:28:07.862 AWST [23229] gnucash@gnucash_rob_professional DETAIL:  Key (id)=(109) already exists.
2020-02-29 12:28:07.862 AWST [23229] gnucash@gnucash_rob_professional STATEMENT:  INSERT INTO slots(obj_guid,name,slot_type,int64_val,timespec_val,guid_val,numeric_val_num,numeric_val_denom) VALUES('f461d9b63d4d4c048e55698f66691312','gncInvoice',9,0,'1970-01-01 00:00:00','8e1a822063824d0e9492bc1688a17131',0,1)
2020-02-29 12:28:08.973 AWST [23229] gnucash@gnucash_rob_professional ERROR:  duplicate key value violates unique constraint "slots_new_pkey"
2020-02-29 12:28:08.973 AWST [23229] gnucash@gnucash_rob_professional DETAIL:  Key (id)=(110) already exists.
2020-02-29 12:28:08.973 AWST [23229] gnucash@gnucash_rob_professional STATEMENT:  INSERT INTO slots(obj_guid,name,slot_type,int64_val,timespec_val,guid_val,numeric_val_num,numeric_val_denom) VALUES('7a0fd3d46a3e4a2f911ec324e5f8ea29','gncInvoice',9,0,'1970-01-01 00:00:00','4ff825c8819e4245b43320bc8dc43dde',0,1)
2020-02-29 12:28:09.795 AWST [23229] gnucash@gnucash_rob_professional ERROR:  duplicate key value violates unique constraint "slots_new_pkey"
2020-02-29 12:28:09.795 AWST [23229] gnucash@gnucash_rob_professional DETAIL:  Key (id)=(111) already exists.
2020-02-29 12:28:09.795 AWST [23229] gnucash@gnucash_rob_professional STATEMENT:  INSERT INTO slots(obj_guid,name,slot_type,int64_val,timespec_val,numeric_val_num,numeric_val_denom) VALUES('d8a5fbd764624b128f2c84ae3c89d67c','credit-note',1,0,'1970-01-01 00:00:00',0,1)
```

It appears that GnuCash is trying to insert duplicate keys into the database.  Why???

**Comment by gjanssens** [comment]
Actually it doesn't look like a GnuCash issue but a db management issue.

Looking closely at the db logs it complains about certain key ids already existing. If you look however at the exact statement that gets executed you'll see the key id is not part of the values being set. The 'id' is an auto-increment field in the database. If you get duplicate ID errors that means the sequences behind these fields don't match the actual data. The most likely cause for this to happen is during the export/import from postgresql 9 to postgresql 12. Did you also migrate the autofield sequences there, or have the import regenerate them ?
You could manually fix it as well by querying for the highest id in each table and adjust the sequence next value for the associated sequence for each of these ids.

**Comment by rob-hills** [comment]
Thanks for your reply.  You were absolutely correct, for some reason it appears that one sequence wasn't restored properly (or at least its nextval wasn't restored correctly): slots_new_id_seq.  As soon as I set that sequence's value to the maximum ID in the slots table, my problems went away.

There were a total of 4 sequences and the other 3 appeared to have the correct values so I'm mystified how this went wrong.

I used the backup/restore process recommended in the official Postgres upgrade notes (pgdump) so I am surprised that one of four sequences was restored with the incorrect value.  I'd have expected either none or all four.

Anyway, apparently not a Gnucash (or Flatpak) problem so I will close this issue

**Comment by asheiduk** [comment]
I've got the same problem when upgrading from Gnucash 2.6.19 to 4.8.

I reproduced the problem by 
- restoring the database
- checking, that the value of the sequence mentioned in the primary key column of `slots` is correct
- starting Gnucash with debug enabled: `gnucash --log gnc.backend.dbi=debug`.
- getting the error message "Unable to save to database"

Now the name of the sequence for the table has changed from `slots_id_seq` to `slots_new_id_seq` and he sequence counter is reset to the default.

The tracefile contains this:
```
[GncDbiSqlConnection::create_table] SQL: CREATE TABLE slots_new(id serial PRIMARY KEY NOT NULL, obj_guid varchar(32) NOT NULL, name varchar(4096) NOT NULL, slot_type integer NOT NULL, int64_val int8, string_val varchar(4096), double_val double precision, timespec_val timestamp without time zone, guid_val varchar(32), numeric_val_num int8, numeric_val_denom int8, gdate_val date)
[GncDbiSqlConnection::execute_nonselect_statement] SQL: INSERT INTO slots_new SELECT * FROM slots
[GncDbiSqlConnection::execute_nonselect_statement] SQL: DROP TABLE slots
[GncDbiSqlConnection::execute_nonselect_statement] SQL: ALTER TABLE slots_new RENAME TO slots
[GncDbiSqlConnection::execute_nonselect_statement] SQL: UPDATE versions SET table_version=4 WHERE table_name='slot
``` 
This explains both the changed name of the sequence (Postgres only renames the table but not the attached sequence) 
*and* the wrong sequence number -- the correct number is never copied before the old sequence is implicitly deleted!

The same thing happens not only for the table `slots` but also for `budget_amounts`, `recurrences` and `taxtable_entries` but these table are empty and hence the wrong sequence number did not matter in my case.

For reference: something like this is missing in the Gnucash code:
```
select setval(pg_get_serial_sequence('slots', 'id'), (select max(id) from slots))
```

**Comment by jralls** [comment]
That sequence of queries is for changing the table's schema. It's legal SQL that works correctly on MySQL and SQLite3.
The field's name is specified as 'id', but Postgres is giving it a secret name that breaks it when the table is renamed. That's a bug in Postgres.

Since it seems unlikely to me that Postgres will fix it I think we can work around it by renaming the existing table to `<table>_old` and creating `<table>`, copying, and dropping `<table>-old`. Does that make sense?


BTW, GnuCash's bug tracker is [here](https://bugs.gnucash.org/). Please open a bug there.

**Comment by asheiduk** [comment]
> That sequence of queries is for changing the table's schema. It's legal SQL that works correctly on MySQL and SQLite3.
It also works in Postgres

> The field's name is specified as 'id', but Postgres is giving it a secret name that breaks it when the table is renamed. That's a bug in Postgres. 
Postgres does not invent column names -- it is explicitly named in the `CREATE TABLE` statement. It also not a problem that the implcitly created sequence is named "non-standard" after the rename -- this works.

What is wrong here is a **logical** error in Gnucash: It copies the data to a new table but does not update the sequence.
How should Postgres know that the sequence number of the old table should also be copied?

>BTW, GnuCash's bug tracker is [here](https://bugs.gnucash.org/). Please open a bug there.
I found this issue (via Google) by searching for the error message. I commented here to help the next one with the same problem. But since Gnucash has its own bugtracker. The friction for creating Yet Another Account For Exactly One Action is to high for me. May others pick up that torch.
